File: cardmarket-frontend/src/api/portfolio.ts
import { apiGet, apiPost, apiPatch, apiDelete } from "./client";

export interface PurchaseItem {
  id: number;
  product_id: number;
  product_name: string;
  product_url: string | null;
  purchase_date: string;
  purchase_price: number;
  quantity: number;
  notes: string | null;
  created_at: string | null;
  current_price: number | null;
  current_total: number | null;
  pl_abs: number | null;
  pl_pct: number | null;
}

export interface PortfolioSummary {
  total_invested: number;
  current_value: number;
  pl_abs: number;
  pl_pct: number | null;
  purchase_count: number;
}

export interface PortfolioHistoryPoint {
  date: string;
  portfolio_value: number;
  invested: number;
}

export interface PurchaseCreatePayload {
  product_id: number;
  purchase_date: string;
  purchase_price: number;
  quantity: number;
  notes?: string;
}

export interface PurchaseUpdatePayload {
  purchase_date?: string;
  purchase_price?: number;
  quantity?: number;
  notes?: string;
}

export const fetchPortfolioPurchases = () =>
  apiGet<PurchaseItem[]>("/portfolio/");

export const fetchPortfolioSummary = () =>
  apiGet<PortfolioSummary>("/portfolio/summary");

export const fetchPortfolioHistory = () =>
  apiGet<PortfolioHistoryPoint[]>("/portfolio/history");

export const addPurchase = (data: PurchaseCreatePayload) =>
  apiPost<PurchaseItem>("/portfolio/", data);

export const updatePurchase = (id: number, data: PurchaseUpdatePayload) =>
  apiPatch<PurchaseItem>(`/portfolio/${id}`, data);

export const deletePurchase = (id: number) =>
  apiDelete(`/portfolio/${id}`);

File: cardmarket-backend/app/schemas/portfolio.py
from datetime import date, datetime
from typing import Optional
from pydantic import BaseModel


class PurchaseCreate(BaseModel):
    product_id: int
    purchase_date: date
    purchase_price: float
    quantity: int = 1
    notes: Optional[str] = None


class PurchaseUpdate(BaseModel):
    purchase_date: Optional[date] = None
    purchase_price: Optional[float] = None
    quantity: Optional[int] = None
    notes: Optional[str] = None


class PurchaseItem(BaseModel):
    id: int
    product_id: int
    product_name: str
    product_url: Optional[str]
    purchase_date: date
    purchase_price: float
    quantity: int
    notes: Optional[str]
    created_at: Optional[datetime]
    current_price: Optional[float]
    current_total: Optional[float]
    pl_abs: Optional[float]
    pl_pct: Optional[float]


class PortfolioSummary(BaseModel):
    total_invested: float
    current_value: float
    pl_abs: float
    pl_pct: Optional[float]
    purchase_count: int


class PortfolioHistoryPoint(BaseModel):
    date: str
    portfolio_value: float
    invested: float
